source/mir/combinatorics/package.d

```diff
@@ -0,0 +1,393 @@
+module mir.combinatorics;
+
+import std.range.primitives: isInputRange, hasLength;
+import std.traits: isAssignable, isFloatingPoint;
+
+/**
+Checks whether we can do basic arithmetic operations, comparisons, modulo and
+assign values to the type.
+*/
+private template isArithmetic(R)
+{
+    enum bool isArithmetic = is(typeof(
+    (inout int = 0)
+    {
+        R r = 1;
+        R test = (r * r / r + r - r) % r;
+        if(r < r && r > r) {}
+    }));
+}
+
+/**
+Checks whether we can do basic arithmetic operations, comparison and modulo
+between two types. R needs to support item assignment of S (it includes S).
+Both R and S need to be arithmetic types themselves.
+*/
+private template isArithmetic(R, S)
+{
+    enum bool isArithmetic = is(typeof(
+    (inout int = 0)
+    {
+        if(isArithmetic!R && isArithmetic!S) {}
+        S s = 1;
+        R r = 1;
+        R test = r * s + r * s;
+        R test2 = r / s + r / s;
+        R test3 = r - s + r - s;
+        R test4 = r % s + r % s;
+        if(r < s && s > r) {}
+        if(s < r && r > s) {}
+    }));
+}
+
+/**
+Computes the $(WEB en.wikipedia.org/wiki/Factorial, factorial) of n -
+also written as `n!`. It is the product of all positive integeres less than or
+equal n. Its generalization to real numbers is called the Gamma function.
+As the factorial grows very quickly, it is recommended to use `BigInt`.
+
+Params:
+    n = factorial number
+
+Returns:
+    Computed factorial
+*/
+auto factorial(R, T)(T n)
```

Comment:
I don't think we need this functions. Only factorial for BigInts is interesting and in the same time proper algorithm is complicated